pkg/registrar/transitgateway.go

```diff
@@ -0,0 +1,268 @@
+		annotations.AddAnnotations(cluster, map[string]string{
+			annotations.NetworkTopologyModeAnnotation: annotations.NetworkTopologyModeNone,
+		})
+		if _, err := r.clusterClient.Save(ctx, cluster); err != nil {
+			logger.Error(err, "Failed to save cluster resource")
+			return err
+		}
+		fallthrough
+
+	case annotations.NetworkTopologyModeNone:
+		// TODO: Handle `None` topology mode
+		logger.Info("Mode currently not handled", "mode", annotations.NetworkTopologyModeNone)
+
+	case annotations.NetworkTopologyModeUserManaged:
+		// TODO: Handle `UserManaged` mode
+		logger.Info("Mode currently not handled", "mode", annotations.NetworkTopologyModeUserManaged)
+
+	case annotations.NetworkTopologyModeGiantSwarmManaged:
+		gatewayID := annotations.GetNetworkTopologyTransitGatewayID(cluster)
+
+		var err error
+		var tgw *types.TransitGateway
+
+		if r.clusterClient.IsManagementCluster(ctx, cluster) {
+			tgw, err = r.getOrCreateTransitGateway(ctx, gatewayID)
+			if err != nil {
+				return err
+			}
+		} else {
+			if gatewayID == "" {
+				// No TGW ID specified so we'll use the one associated with the MC
+				mc, err := r.clusterClient.GetManagementCluster(ctx)
+				if err != nil {
+					logger.Error(err, "Failed to get management cluster")
+					return err
+				}
+
+				gatewayID = annotations.GetNetworkTopologyTransitGatewayID(mc)
+				if gatewayID == "" {
+					err = fmt.Errorf("management cluster doesn't have a TGW specified")
+					logger.Error(err, "The Management Cluster doesn't have a Transit Gateway ID specified")
+					return err
+				}
+			}
+
+			tgw, err = r.getTransitGateway(ctx, gatewayID)
+			if err != nil {
+				return err
+			}
+			if tgw == nil {
+				err = fmt.Errorf("failed to find TransitGateway for provided ID")
+				logger.Error(err, "No TransitGateway found for ID provided on annotations", "transitGatewayID", gatewayID)
+				return err
+			}
+		}
+
+		// Ensure TGW ID is saved back to the current cluster
+		annotations.SetNetworkTopologyTransitGatewayID(cluster, *tgw.TransitGatewayId)
+		if cluster, err = r.clusterClient.Save(ctx, cluster); err != nil {
+			logger.Error(err, "Failed to save cluster resource")
+			return err
+		}
+
+		if err := r.attachTransitGateway(ctx, tgw.TransitGatewayId, cluster); err != nil {
+			logger.Error(err, "Failed to attach transit gateway to cluster VPC")
+			return err
+		}
+
+	default:
+		err := fmt.Errorf("invalid NetworkTopologyMode value")
+		logger.Error(err, "Unexpected NetworkTopologyMode annotation value found on cluster", "value", val)
+		return err
+	}
+
+	logger.Info("Done Registering TransitGateway")
+	return nil
+}
+
+func (r *TransitGateway) Unregister(ctx context.Context, cluster *capi.Cluster) error {
+	logger := r.getLogger(ctx)
+
+	if annotations.IsNetworkTopologyModeGiantSwarmManaged(cluster) {
+		logger.Info("Network topology mode is GiantSwarmManaged")
+
+		// TODO: Check if TGW needs deleting
+
```

Comment:
we will need to detach the VPC in case of WC 
if its MC than probably detach all VPCs and delete TGW

Reply:
Does a TGW need detaching before the VPC is deleted or will it just happen during delete?

Reply:
Added anyway, can't hurt to detach first :)